document modchannels (#15)
minetest/minetest_docs: doc/modchannels.adoc

= Modchannels

> allows ssm(server side mods) comunication to csm(client side mods) and vice versa.

Reviewer: ```suggestion
Allows server side mods (SSMs) communication to client side mods (CSMs) and vice versa.
```

> == minetest.x api

Reviewer: ```suggestion
== Functions
```

= Modchannels

> allows server side mods(SSMs) comunication to client side mods(CSMs) and vice versa.

Reviewer: ```suggestion
Allows server side mods (SSMs) communication to client side mods (CSMs) and vice versa.
```

Author: https://github.com/minetest/minetest_docs/pull/15/commits/fe0524f994a3c99b6f18e8bda4481e7592ee5419

== Server Side API

=== Functions

==== minetest.mod_channel_join(channel_name)

* `channel_name`: string, modchannel name
* returns ref for use with <<Methods>>
* creates channel if not existing and joins the server on that channel

> * listen for messages on this channel with `minetest.register_on_modchannel_message`

Reviewer: Why are all of these bullet points, and in the same paragraph? I suppose that it does do all of these things, but this feels like a very "not plain English" way of describing it.

Maybe something like:
```asciidoc
==== `minetest.mod_channel_join(channel_name)
* `channel_name`: `string`

Returns a new mod channel object with <<Methods>> for interacting with the channel.

This will join an existing channel named `channel_name` if it exists, or create (and then join) it as a new channel if it does not.
```

Author: https://github.com/minetest/minetest_docs/pull/15/commits/ef9ef4d1b5a87999062fe2a04c838f3a1b97b433

==== minetest.register_on_modchannel_message(function(channel_name, sender, message))

* `channel_name`: string, modchannel name (already joined)
* `sender`: string, empty if from a ssm, player name if from a client
* `message`: string, message

=== Methods

> 
> ==== leave()

Reviewer: I think that showing the intended call is useful here.

```suggestion
NOTE: `obj` here means the object returned by `minetest.mod_channel_join`.

==== obj:leave()
```

Author: https://github.com/minetest/minetest_docs/pull/15/commits/fe0524f994a3c99b6f18e8bda4481e7592ee5419

= Modchannels

Allows server side mods (SSMs) communication to client side mods (CSMs) and vice versa.

== Server Side API

=== Functions

==== minetest.mod_channel_join(channel_name)

* `channel_name`: string, modchannel name

> Returns a object for use with <<Methods>>. Creates the channel if it does not exist and joins the channel.

Reviewer: ```suggestion
Returns an object for use with <<Methods>>. Creates the channel if it does not exist and joins the channel.
```

Author: https://github.com/minetest/minetest_docs/pull/15/commits/3940703fc9579f4c929b80eb614a62b11b20dbbf

= Modchannels

Allows server side mods (SSMs) communication to client side mods (CSMs) and vice versa.

== Server Side API

=== Functions

==== minetest.mod_channel_join(channel_name)

* `channel_name`: string, modchannel name

Returns a object for use with <<Methods>>. Creates the channel if it does not exist and joins the channel.

==== minetest.register_on_modchannel_message(function(channel_name, sender, message))

* `channel_name`: string, modchannel name (already joined)
* `sender`: string, empty if from a ssm, player name if from a client

> * `message`: string, message

Reviewer: This needs an explanation for what it does/is useful for.

Author: https://github.com/minetest/minetest_docs/pull/15/commits/3940703fc9579f4c929b80eb614a62b11b20dbbf

=== Methods

NOTE: `channel` here means the object returned by `minetest.mod_channel_join`.

==== channel:leave()

The server will leave the channel, meaning no more messages from this channel on `minetest.register_on_modchannel_message`

> TIP: set the channel to nil afterwords to free resources

Reviewer: ```suggestion
TIP: Set the channel to `nil` afterwords to free resources.
```

Author: https://github.com/minetest/minetest_docs/pull/15/commits/3940703fc9579f4c929b80eb614a62b11b20dbbf

==== channel:is_writeable()

> * returns boolean, true if the channel is writeable

Reviewer: ```suggestion
* Returns a `bool`: `true` if the channel is writable, `false` if it's not.
```

Author: https://github.com/minetest/minetest_docs/pull/15/commits/3940703fc9579f4c929b80eb614a62b11b20dbbf

==== channel:send_all(message)

* `message`: string, limited to 65535 bytes

Sends to all SSMs and CSMs on the channel.

> TIP: The message will drop if channel is not writable or invalid

Reviewer: ```suggestion
CAUTION: The message will not send if channel is not writable or invalid.
```

Author: https://github.com/minetest/minetest_docs/pull/15/commits/3940703fc9579f4c929b80eb614a62b11b20dbbf

== Client Side API

=== Functions

==== minetest.mod_channel_join(channel_name)

* `channel_name`: string, modchannel name

Returns a object for use with <<Methods>>. Creates the channel if it does not exist and joins the channel.

==== minetest.register_on_modchannel_message(function(channel_name, sender, message))

* `channel_name`: string, modchannel name (already joined and recieved acknowledgement)
* `sender`: string, empty if from a ssm, player name if from a client

> * `message`: string, message

Reviewer: Similar to the one above, can be said to be equivalent.

Author: https://github.com/minetest/minetest_docs/pull/15/commits/3940703fc9579f4c929b80eb614a62b11b20dbbf

==== minetest.register_on_modchannel_signal(function(channel_name, signal))
* `channel_name`: string, channel name that the signal has come in on
* `signal`: integer, 0 - 5
[start=0]
. join_ok
. join_failed
. leave_ok
. leave_failed
. event_on_not_joined_channel
. state_changed

=== Methods

NOTE: `channel` here means the object returned by `minetest.mod_channel_join`.

==== channel:leave()

> The server will leave the channel, meaning no more messages from this channel on `minetest.register_on_modchannel_message`

Reviewer: ```suggestion
The client will leave the channel, meaning no more messages from this channel on `minetest.register_on_modchannel_message`
```

Author: https://github.com/minetest/minetest_docs/pull/15/commits/3940703fc9579f4c929b80eb614a62b11b20dbbf

> TIP: set the channel to nil afterwords to free resources

Reviewer: ```suggestion
TIP: Set the channel to `nil` afterwords to free resources.
```

Author: https://github.com/minetest/minetest_docs/pull/15/commits/3940703fc9579f4c929b80eb614a62b11b20dbbf

==== channel:is_writeable()

> * returns boolean, true if the channel is writeable

Reviewer: ```suggestion
* Returns a `bool`: `true` if the channel is writable, `false` if it's not.
```

Author: https://github.com/minetest/minetest_docs/pull/15/commits/3940703fc9579f4c929b80eb614a62b11b20dbbf

==== channel:send_all(message)

* `message`: string, limited to 65535 bytes

Sends to all SSMs and CSMs on the channel.

> TIP: The message will drop if channel is not writable or invalid

Reviewer: ```suggestion
CAUTION: The message will not send if the channel is not writable or invalid.
```

Author: https://github.com/minetest/minetest_docs/pull/15/commits/3940703fc9579f4c929b80eb614a62b11b20dbbf

= Modchannels

Allows server side mods (SSMs) communication to client side mods (CSMs) and vice versa.

== Server Side API

=== Functions

==== minetest.mod_channel_join(channel_name)

* `channel_name`: string, modchannel name

Returns an object for use with <<Methods>>. Creates the channel if it does not exist and joins the channel.

==== minetest.register_on_modchannel_message(function(channel_name, sender, message))

* `channel_name`: string, modchannel name (already joined)
* `sender`: string, empty if from a SSM, player name if from a client
* `message`: string, message

Used for handling messages received from the client.

=== Methods

NOTE: `channel` here means the object returned by `minetest.mod_channel_join`.

==== channel:leave()

The server will leave the channel, meaning no more messages from this channel on `minetest.register_on_modchannel_message`

TIP: Set the channel to `nil`` afterwords to free resources

==== channel:is_writeable()

* Returns `bool`: `true` true if the channel is writeable, `false` if it's not.

==== channel:send_all(message)

* `message`: string, limited to 65535 bytes

Sends to all SSMs and CSMs on the channel.

> Caution: The message will not if channel is not writable or invalid.

Reviewer: ```suggestion
CAUTION: The message will not if channel is not writable or invalid.
```

== Client Side API

=== Functions

==== minetest.mod_channel_join(channel_name)

* `channel_name`: string, modchannel name

Returns an object for use with <<Methods>>. Creates the channel if it does not exist and joins the channel. Is equivalent to the the server side function.

==== minetest.register_on_modchannel_message(function(channel_name, sender, message))

* `channel_name`: string, modchannel name (already joined and recieved acknowledgement)
* `sender`: string, empty if from a SSM, player name if from a client
* `message`: string, message

Used for handling messages received from the client. Is equivalent to the the server side function.

> ==== minetest.register_on_modchannel_signal(function(channel_name, signal))
> * `channel_name`: string, channel name that the signal has come in on
> * `signal`: integer, 0 - 5
> [start=0]
> . join_ok
> . join_failed
> . leave_ok
> . leave_failed
> . event_on_not_joined_channel
> . state_changed
> 

Reviewer: ```suggestion
==== minetest.register_on_modchannel_signal(function(channel_name, signal))
* `channel_name`: string, channel name that the signal has come in on
* `signal`: integer, 0 - 5
[start=0]
. join_ok
. join_failed
. leave_ok
. leave_failed
. event_on_not_joined_channel
. state_changed

Used to handle signals generated by the mod channel system.
```

=== Methods

NOTE: `channel` here means the object returned by `minetest.mod_channel_join`.

==== channel:leave()

The client will leave the channel, meaning no more messages from this channel on `minetest.register_on_modchannel_message`

> TIP: Set the channel to `nil`` afterwords to free resources

Reviewer: ```suggestion
TIP: Set the channel to `nil` afterwords to free resources
```

==== channel:is_writeable()

* Returns `bool`: `true` true if the channel is writeable, `false` if it's not.

==== channel:send_all(message)

* `message`: string, limited to 65535 bytes

Sends to all SSMs and CSMs on the channel.

> Caution: The message will not if channel is not writable or invalid.

Reviewer: ```suggestion
CAUTION: The message will not if channel is not writable or invalid.
```

==== channel:send_all(message)

* `message`: string, limited to 65535 bytes

Sends to all SSMs and CSMs on the channel.

CAUTION: The message will not if channel is not writable or invalid.

== Client Side API

=== Functions

==== minetest.mod_channel_join(channel_name)

* `channel_name`: string, modchannel name

Returns an object for use with <<Methods>>. Creates the channel if it does not exist and joins the channel. Is equivalent to the the server side function.

==== minetest.register_on_modchannel_message(function(channel_name, sender, message))

* `channel_name`: string, modchannel name (already joined and recieved acknowledgement)
* `sender`: string, empty if from a SSM, player name if from a client
* `message`: string, message

Used for handling messages received from the client. Is equivalent to the the server side function.

==== minetest.register_on_modchannel_signal(function(channel_name, signal))
* `channel_name`: string, channel name that the signal has come in on
* `signal`: integer, 0 - 5
[start=0]
. join_ok
. join_failed
. leave_ok
. leave_failed
. event_on_not_joined_channel
. state_changed

Used to handle signals generated by the mod channel system.

> === Methods

Reviewer: Last thing
```suggestion

=== Methods
```
